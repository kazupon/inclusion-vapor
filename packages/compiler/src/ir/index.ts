export * from './svelte'
export * from './nodes'
export * from './component'
