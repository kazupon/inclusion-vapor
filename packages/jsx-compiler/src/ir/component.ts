// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `unplugin-vue-jsx-vapor`
// Author: zhiyuanzmj (https://github.com/zhiyuanzmj) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/unplugin/unplugin-vue-jsx-vapor
// Code url: https://github.com/unplugin/unplugin-vue-jsx-vapor/tree/main/src/core/compiler

import type { IRDynamicPropsKind } from '@vue-vapor/compiler-vapor'

import type { SimpleExpressionNode } from '@vue-vapor/compiler-dom'
import type { DirectiveTransformResult } from '../transforms/index.ts'
import type { BlockIRNode, IRFor } from './nodes.ts'

// props
export interface IRProp extends Omit<DirectiveTransformResult, 'value'> {
  values: SimpleExpressionNode[]
}

export type IRPropsStatic = IRProp[]
export interface IRPropsDynamicExpression {
  kind: IRDynamicPropsKind.EXPRESSION
  value: SimpleExpressionNode
  handler?: boolean
}
export interface IRPropsDynamicAttribute extends IRProp {
  kind: IRDynamicPropsKind.ATTRIBUTE
}
export type IRProps = IRPropsStatic | IRPropsDynamicAttribute | IRPropsDynamicExpression

// slots
export interface SlotBlockIRNode extends BlockIRNode {
  props?: SimpleExpressionNode
}

export enum IRSlotType {
  STATIC,
  DYNAMIC,
  LOOP,
  CONDITIONAL,
  EXPRESSION
}

export interface IRSlotsStatic {
  slotType: IRSlotType.STATIC
  slots: Record<string, SlotBlockIRNode>
}
export interface IRSlotDynamicBasic {
  slotType: IRSlotType.DYNAMIC
  name: SimpleExpressionNode
  fn: SlotBlockIRNode
}
export interface IRSlotDynamicLoop {
  slotType: IRSlotType.LOOP
  name: SimpleExpressionNode
  fn: SlotBlockIRNode
  loop: IRFor
}
export interface IRSlotDynamicConditional {
  slotType: IRSlotType.CONDITIONAL
  condition: SimpleExpressionNode
  positive: IRSlotDynamicBasic
  negative?: IRSlotDynamicBasic | IRSlotDynamicConditional
}
export interface IRSlotsExpression {
  slotType: IRSlotType.EXPRESSION
  slots: SimpleExpressionNode
}

export type IRSlotDynamic = IRSlotDynamicBasic | IRSlotDynamicLoop | IRSlotDynamicConditional
export type IRSlots = IRSlotsStatic | IRSlotDynamic | IRSlotsExpression
