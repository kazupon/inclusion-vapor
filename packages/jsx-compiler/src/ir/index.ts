// SPDX-License-Identifier: MIT
// Author: kazuya kawaguchi (a.k.a. kazupon)

export * from './nodes'
export * from './component'
export * from './jsx'
