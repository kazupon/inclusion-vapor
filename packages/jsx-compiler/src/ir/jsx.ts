// SPDX-License-Identifier: MIT
// Author: kazuya kawaguchi (a.k.a. kazupon)

export type {
  CallExpression as BabelCallExpression,
  Expression as BabelExpression,
  Node as BabelNode,
  Program as BabelProgram,
  SourceLocation as BabelSourceLocation,
  JSXAttribute,
  JSXElement,
  JSXExpressionContainer,
  JSXFragment,
  JSXSpreadAttribute,
  JSXText
} from '@babel/types'
