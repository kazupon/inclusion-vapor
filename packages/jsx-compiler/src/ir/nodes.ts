// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `unplugin-vue-jsx-vapor`
// Author: zhiyuanzmj (https://github.com/zhiyuanzmj) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/unplugin/unplugin-vue-jsx-vapor
// Code url: https://github.com/unplugin/unplugin-vue-jsx-vapor/tree/main/src/core/compiler

import type { IRNodeTypes, IRDynamicInfo } from '@vue-vapor/compiler-vapor'

import type {
  BindingTypes,
  CompoundExpressionNode,
  DirectiveNode,
  SimpleExpressionNode
} from '@vue-vapor/compiler-dom'
import type { JSXFragment, BabelNode } from './jsx'
import type { IRProp, IRProps, IRSlots } from './component'
import type { Overwrite } from '../types'

export interface BaseIRNode {
  type: IRNodeTypes
}

export interface RootNode {
  type: IRNodeTypes.ROOT
  source: string
  children: JSXFragment['children']
  helpers: Set<symbol>
  components: string[]
  directives: string[]
  // loc: JSXElement['loc']
  // hoists: (JSChildNode | null)[]
  // imports: ImportItem[]
  // cached: (CacheExpression | null)[]
  temps: number
  ssrHelpers?: symbol[]
  // codegenNode?: TemplateChildNode | JSChildNode | BlockStatement
  transformed?: boolean
}

export interface BlockIRNode extends BaseIRNode {
  type: IRNodeTypes.BLOCK
  node: RootNode | BabelNode
  dynamic: IRDynamicInfo
  effect: IREffect[]
  operation: OperationNode[]
  returns: number[]
}

export interface RootIRNode {
  type: IRNodeTypes.ROOT
  node: RootNode
  source: string
  template: string[]
  component: Set<string>
  directive: Set<string>
  block: BlockIRNode
}

export interface IfIRNode extends BaseIRNode {
  type: IRNodeTypes.IF
  id: number
  condition: SimpleExpressionNode
  positive: BlockIRNode
  negative?: BlockIRNode | IfIRNode
  once?: boolean
}

export interface IRFor {
  source: SimpleExpressionNode
  value?: SimpleExpressionNode
  key?: SimpleExpressionNode
  index?: SimpleExpressionNode
}

export interface ForIRNode extends BaseIRNode, IRFor {
  type: IRNodeTypes.FOR
  id: number
  keyProp?: SimpleExpressionNode
  render: BlockIRNode
  once: boolean
}

export interface SetPropIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_PROP
  element: number
  prop: IRProp
}

export interface SetDynamicPropsIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_DYNAMIC_PROPS
  element: number
  props: IRProps[]
}

export interface SetDynamicEventsIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_DYNAMIC_EVENTS
  element: number
  event: SimpleExpressionNode
}

export interface SetTextIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_TEXT
  element: number
  values: SimpleExpressionNode[]
}

export type KeyOverride = [find: string, replacement: string]
export interface SetEventIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_EVENT
  element: number
  key: SimpleExpressionNode
  value?: SimpleExpressionNode
  modifiers: {
    // modifiers for addEventListener() options, e.g. .passive & .capture
    options: string[]
    // modifiers that needs runtime guards, withKeys
    keys: string[]
    // modifiers that needs runtime guards, withModifiers
    nonKeys: string[]
  }
  keyOverride?: KeyOverride
  delegate: boolean
  /** Whether it's in effect */
  effect: boolean
}

export interface SetHtmlIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_HTML
  element: number
  value: SimpleExpressionNode
}

export interface SetTemplateRefIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_TEMPLATE_REF
  element: number
  value: SimpleExpressionNode
  refFor: boolean
  effect: boolean
}

export interface SetModelValueIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_MODEL_VALUE
  element: number
  key: SimpleExpressionNode
  value: SimpleExpressionNode
  bindingType?: BindingTypes
  isComponent: boolean
}

export interface CreateTextNodeIRNode extends BaseIRNode {
  type: IRNodeTypes.CREATE_TEXT_NODE
  id: number
  values: SimpleExpressionNode[]
  effect: boolean
}

export interface InsertNodeIRNode extends BaseIRNode {
  type: IRNodeTypes.INSERT_NODE
  elements: number[]
  parent: number
  anchor?: number
}

export interface PrependNodeIRNode extends BaseIRNode {
  type: IRNodeTypes.PREPEND_NODE
  elements: number[]
  parent: number
}

export interface WithDirectiveIRNode extends BaseIRNode {
  type: IRNodeTypes.WITH_DIRECTIVE
  element: number
  dir: VaporDirectiveNode
  name: string
  builtin?: boolean
  asset?: boolean
}

export interface CreateComponentIRNode extends BaseIRNode {
  type: IRNodeTypes.CREATE_COMPONENT_NODE
  id: number
  tag: string
  props: IRProps[]
  slots: IRSlots[]
  asset: boolean
  root: boolean
  once: boolean
}

export interface DeclareOldRefIRNode extends BaseIRNode {
  type: IRNodeTypes.DECLARE_OLD_REF
  id: number
}

export interface SlotOutletIRNode extends BaseIRNode {
  type: IRNodeTypes.SLOT_OUTLET_NODE
  id: number
  name: SimpleExpressionNode
  props: IRProps[]
  fallback?: BlockIRNode
}

export type IRNode = OperationNode | RootIRNode
export type OperationNode =
  | SetPropIRNode
  | SetDynamicPropsIRNode
  | SetTextIRNode
  | SetEventIRNode
  | SetDynamicEventsIRNode
  | SetHtmlIRNode
  | SetTemplateRefIRNode
  | SetModelValueIRNode
  | CreateTextNodeIRNode
  | InsertNodeIRNode
  | PrependNodeIRNode
  | WithDirectiveIRNode
  | IfIRNode
  | ForIRNode
  | CreateComponentIRNode
  | DeclareOldRefIRNode
  | SlotOutletIRNode

export interface IREffect {
  expressions: SimpleExpressionNode[]
  operations: OperationNode[]
}

export type VaporDirectiveNode = Overwrite<
  DirectiveNode,
  {
    exp: Exclude<DirectiveNode['exp'], CompoundExpressionNode>
    arg: Exclude<DirectiveNode['arg'], CompoundExpressionNode>
  }
>

export type { IRDynamicInfo } from '@vue-vapor/compiler-vapor'
export { IRNodeTypes, DynamicFlag, IRDynamicPropsKind } from '@vue-vapor/compiler-vapor'
