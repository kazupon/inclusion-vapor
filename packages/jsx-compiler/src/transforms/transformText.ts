// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `unplugin-vue-jsx-vapor`
// Author: zhiyuanzmj (https://github.com/zhiyuanzmj) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/unplugin/unplugin-vue-jsx-vapor
// Code url: https://github.com/unplugin/unplugin-vue-jsx-vapor/tree/main/src/core/compiler

import { DynamicFlag, IRNodeTypes } from '../ir/index.ts'
// import { processConditionalExpression, processLogicalExpression } from './vIf'
// import { processMapCallExpression } from './vFor'
import {
  getLiteralExpressionValue,
  isComponentNode,
  isConstantExpression,
  isMapCallExpression,
  resolveExpression
} from './utils.ts'

import type {
  BabelCallExpression,
  BabelNode,
  BlockIRNode,
  JSXElement,
  JSXExpressionContainer,
  JSXText,
  RootNode
} from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { NodeTransform } from './types.ts'

type TextLike = JSXText | JSXExpressionContainer

const seen = new WeakMap<
  TransformContext<RootNode>,
  WeakSet<TextLike | BlockIRNode['node'] | RootNode>
>()

export const transformText: NodeTransform = (node, context) => {
  if (!seen.has(context.root)) {
    seen.set(context.root, new WeakSet())
  }

  if (seen.get(context.root)!.has(node)) {
    context.dynamic.flags |= DynamicFlag.NON_TEMPLATE
    return
  }

  if (
    node.type === 'JSXElement' &&
    !isComponentNode(node) &&
    // @ts-expect-error -- FIXME
    isAllTextLike(node.children) // eslint-disable-line @typescript-eslint/no-unsafe-argument
  ) {
    processTextLikeContainer(
      // @ts-expect-error -- FIXME
      node.children, // eslint-disable-line @typescript-eslint/no-unsafe-argument
      context as TransformContext<JSXElement>
    )
  } else if (node.type === 'JSXExpressionContainer') {
    switch (node.expression.type) {
      // case 'ConditionalExpression': {
      //   return processConditionalExpression(node.expression, context)
      // }
      // case 'LogicalExpression': {
      //   return processLogicalExpression(node.expression, context)
      // }
      case 'CallExpression': {
        if (isMapCallExpression(node.expression)) {
          // return processMapCallExpression(node.expression, context)
        } else {
          processCallExpression(node.expression, context)
        }

        break
      }
      default: {
        processTextLike(context as TransformContext<JSXExpressionContainer>)
      }
    }
  } else if (node.type === 'JSXText') {
    context.template += node.value
  }
}

function processTextLike(context: TransformContext<JSXExpressionContainer>) {
  const nexts = context.parent!.node.children?.slice(context.index)
  const idx = nexts.findIndex(n => !isTextLike(n))
  const nodes = (idx > -1 ? nexts.slice(0, idx) : nexts) as Array<TextLike>

  const id = context.reference()
  const values = nodes.map(node => createTextLikeExpression(node, context))

  context.dynamic.flags |= DynamicFlag.INSERT | DynamicFlag.NON_TEMPLATE

  context.registerOperation({
    type: IRNodeTypes.CREATE_TEXT_NODE,
    id,
    values,
    // eslint-disable-next-line unicorn/no-array-callback-reference
    effect: !values.every(isConstantExpression) && !context.inVOnce
  })
}

function processTextLikeContainer(children: TextLike[], context: TransformContext<JSXElement>) {
  const values = children.map(child => createTextLikeExpression(child, context))
  const literals = values.map(getLiteralExpressionValue) // eslint-disable-line unicorn/no-array-callback-reference
  if (literals.every(l => l != undefined)) {
    context.childrenTemplate = literals.map(String)
  } else {
    context.registerEffect(values, {
      type: IRNodeTypes.SET_TEXT,
      element: context.reference(),
      values
    })
  }
}

function createTextLikeExpression(node: TextLike, context: TransformContext) {
  seen.get(context.root)!.add(node)
  return resolveExpression(node, context)
}

function isAllTextLike(children: BabelNode[]): children is TextLike[] {
  return (
    children.length > 0 &&
    // eslint-disable-next-line unicorn/no-array-callback-reference
    children.every(isTextLike) &&
    // at least one an interpolation
    children.some(n => n.type === 'JSXExpressionContainer')
  )
}

function isTextLike(node: BabelNode): node is TextLike {
  return (
    (node.type === 'JSXExpressionContainer' &&
      !(
        node.expression.type === 'ConditionalExpression' ||
        node.expression.type === 'LogicalExpression'
      ) &&
      node.expression.type !== 'CallExpression') ||
    node.type === 'JSXText'
  )
}

function processCallExpression(node: BabelCallExpression, context: TransformContext) {
  context.dynamic.flags |= DynamicFlag.NON_TEMPLATE | DynamicFlag.INSERT
  const root = context.root === context.parent && context.parent.node.children.length === 1
  const tag = `() => ${context.ir.source.slice(node.start!, node.end!)}`

  context.registerOperation({
    type: IRNodeTypes.CREATE_COMPONENT_NODE,
    id: context.reference(),
    tag,
    props: [],
    asset: false,
    root,
    slots: context.slots,
    once: context.inVOnce
  })
}
