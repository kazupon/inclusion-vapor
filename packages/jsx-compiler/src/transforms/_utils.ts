import { parse } from '@babel/parser'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, JSXElement, JSXFragment } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const {
      body: [statement]
    } = parse(source, {
      sourceType: 'module',
      plugins: ['jsx']
    }).program
    let children!: JSXElement[] | JSXFragment['children']
    if (statement.type === 'ExpressionStatement') {
      children =
        statement.expression.type === 'JSXFragment'
          ? statement.expression.children
          : (statement.expression.type === 'JSXElement'
            ? [statement.expression]
            : [])
    }

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children,
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
