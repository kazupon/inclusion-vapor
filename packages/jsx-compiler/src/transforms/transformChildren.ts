// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `unplugin-vue-jsx-vapor`
// Author: zhiyuanzmj (https://github.com/zhiyuanzmj) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/unplugin/unplugin-vue-jsx-vapor
// Code url: https://github.com/unplugin/unplugin-vue-jsx-vapor/tree/main/src/core/compiler

import { isComponentNode } from './utils'
import { IRNodeTypes, DynamicFlag } from '../ir'
import { transformNode } from '../transform'

import type { BabelNode, IRDynamicInfo } from '../ir'
import type { NodeTransform } from './types'
import type { TransformContext } from './context'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT || node.type === 'JSXFragment' || isComponentNode(node)

  if (node.type !== 'JSXElement' && !isFragment) {
    return
  }

  ;[...node.children].forEach((child, index) => {
    if (child.type === 'JSXText' && !child.value.trim()) {
      child.value = ' '
      if (!index) {
        node.children.splice(0, 1)
      } else if (index === node.children.length) {
        node.children.splice(-1, 1)
      }
    }
  })

  for (const [i, child] of node.children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }

  if (!isFragment) {
    processDynamicChildren(context as TransformContext<BabelNode>)
  }
}

function processDynamicChildren(context: TransformContext<BabelNode>) {
  let prevDynamics: IRDynamicInfo[] = []
  let hasStaticTemplate = false
  const children = context.dynamic.children

  for (const [index, child] of children.entries()) {
    if (child.flags & DynamicFlag.INSERT) {
      prevDynamics.push(child)
    }

    if (!(child.flags & DynamicFlag.NON_TEMPLATE)) {
      if (prevDynamics.length > 0) {
        if (hasStaticTemplate) {
          context.childrenTemplate[index - prevDynamics.length] = `<!>`

          prevDynamics[0].flags -= DynamicFlag.NON_TEMPLATE
          const anchor = (prevDynamics[0].anchor = context.increaseId())

          context.registerOperation({
            type: IRNodeTypes.INSERT_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference(),
            anchor
          })
        } else {
          context.registerOperation({
            type: IRNodeTypes.PREPEND_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference()
          })
        }
        prevDynamics = []
      }
      hasStaticTemplate = true
    }
  }

  if (prevDynamics.length > 0) {
    context.registerOperation({
      type: IRNodeTypes.INSERT_NODE,
      elements: prevDynamics.map(child => child.id!),
      parent: context.reference()
    })
  }
}
