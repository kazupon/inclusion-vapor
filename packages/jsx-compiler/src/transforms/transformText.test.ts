import { expect, test } from 'vitest'
import { makeCompile, DEFAULT_VAPOR_COMPILER_OPTIONS } from './_utils'
import { transformText } from './transformText'
import { transformElement } from './transformElement'
import { transformChildren } from './transformChildren'
import { IRNodeTypes } from '../ir'
import { compile as vaporCompile } from '@vue-vapor/compiler-vapor'

const compileWithTextTransform = makeCompile({
  nodeTransforms: [transformChildren, transformElement, transformText]
})

test('no consecutive text', () => {
  const { code, ir, vaporHelpers } = compileWithTextTransform('<>{ "hello world" }</>')
  const expectedResult = vaporCompile('{{ "hello world" }}', DEFAULT_VAPOR_COMPILER_OPTIONS)
  expect(code).toMatchSnapshot()
  expect(code).toEqual(expectedResult.code)
  expect(vaporHelpers).contains.all.keys('createTextNode')
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.CREATE_TEXT_NODE,
      id: 0,
      values: [
        {
          type: IRNodeTypes.SET_TEXT,
          content: '"hello world"',
          isStatic: false
        }
      ],
      effect: false
    }
  ])
})

test('consecutive text', () => {
  const { code, ir, vaporHelpers } = compileWithTextTransform('<>{ msg }</>')
  const expectedResult = vaporCompile('{{ msg }}', DEFAULT_VAPOR_COMPILER_OPTIONS)
  expect(code).toMatchSnapshot()
  expect(code).toEqual(expectedResult.code)
  expect(vaporHelpers).contains.all.keys('createTextNode')
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.CREATE_TEXT_NODE,
      id: 0,
      values: [
        {
          type: IRNodeTypes.SET_TEXT,
          content: 'msg',
          isStatic: false
        }
      ],
      effect: true
    }
  ])
})
