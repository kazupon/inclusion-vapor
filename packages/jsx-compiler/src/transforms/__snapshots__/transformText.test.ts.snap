// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`consecutive text 1`] = `
"import { createTextNode as _createTextNode } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createTextNode(() => [_ctx.msg])
  return n0
}"
`;

exports[`no consecutive text 1`] = `
"import { createTextNode as _createTextNode } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createTextNode(["hello world"])
  return n0
}"
`;
